stellargraph/layer/knowledge_graph.py

```diff
@@ -670,6 +671,200 @@ def __init__(
         )
 
 
+class RotHEScoring(Layer, KGScore):
+    def __init__(self, hyperbolic):
+        self._hyperbolic = hyperbolic
+        if self._hyperbolic:
+            self._convert = lambda c, v: poincare_ball_exp(c, None, v)
+            self._add = poincare_ball_mobius_add
+            self._squared_distance = lambda c, v, w: tf.square(
+                poincare_ball_distance(c, v, w)
+            )
+        else:
+            self._convert = lambda _c, v: v
+            self._add = lambda _c, v, w: v + w
+            self._squared_distance = lambda _c, v, w: tf.reduce_sum(
+                tf.math.squared_difference(v, w), axis=-1
+            )
+
+        super().__init__()
+
+    def embeddings(
+        self, num_nodes, num_edge_types, dimension, initializer, regularizer
+    ):
+        if dimension % 2 != 0:
+            raise ValueError(
+                f"embedding_dimension: expected an even integer, found {dimension}"
+            )
+
+        def embed(count, dim=dimension):
+            return Embedding(
+                count,
+                dim,
+                embeddings_initializer=initializer,
+                embeddings_regularizer=regularizer,
+            )
+
+        nodes = [embed(num_nodes), embed(num_nodes, 1)]
+        edge_types = [embed(num_edge_types), embed(num_edge_types, dimension // 2)]
+        return nodes, edge_types
+
+    def build(self, input_shapes):
+        if self._hyperbolic:
+            self.curvature_prime = self.add_weight(shape=(1,), name="curvature_prime")
+        else:
+            self.curvature_prime = None
+
+        super().build(input_shapes)
+
+    def _curvature(self, expand=1):
+        if not self._hyperbolic:
+            return tf.constant([0.0])
+
+        return tf.math.softplus(self.curvature_prime)
```

Comment:
should this check if `curvature_prime` is None? (i.e. `build` hasn't been called)

Reply:
Sounds good.